File: coding-problems/js/e-136-single-number.js
var singleNumber = function (nums) {
    let set_nums = new Set();

    for (num of nums) {
        if (set_nums.has(num)) {
            set_nums.delete(num);
        } else {
            set_nums.add(num);
        }
    }

    return [...set_nums][0];
};

console.log(singleNumber([2, 2, 1]));
console.log(singleNumber([4, 1, 1, 2, 2]));

File: coding-problems/ruby/e-136-single-number.rb
require 'set'

def single_number(nums)
    
    # set_nums = Set[nums.shift]
    set_nums = Set[]

    nums.each do |ele|
        set_nums === ele ? set_nums.delete(ele) : set_nums.add(ele)
    end
    
    return set_nums.to_a()[0]
end

p single_number([2,1,2])
p single_number([4,1,2,1,2])

def single_number_math(nums)
    nums.to_set().to_a().sum * 2 - nums.sum
end

p single_number_math([2,1,2])
p single_number_math([4,1,2,1,2])